File: integral_test.py
import numpy as np
from scipy.integrate import quad, quadrature
from numpy import trapz, cumsum  # 从 numpy 导入 trapz，用 cumsum 模拟 cumtrapz
import matplotlib.pyplot as plt

# 设置中文字体
plt.rcParams['font.sans-serif'] = ['SimHei']  # 使用黑体
plt.rcParams['axes.unicode_minus'] = False    # 解决负号显示问题

# 参数设置
n_segments = 4  # 分段数，可以调整来控制精度
a, b = -1, 1      # 积分区间
x_vals = np.linspace(a, b, 100)  # 用于绘制的采样点

# 定义被积函数
f = lambda x: np.pi * np.cos(np.pi * x)

# 定义原函数 (理论积分结果)
F_exact = np.sin(np.pi * x_vals)

# 初始化积分结果
F_integral = np.zeros(len(x_vals))
F_custom_trapz = np.zeros(len(x_vals))
F_scipy_trapz = np.zeros(len(x_vals))
F_cumtrapz = np.zeros(len(x_vals))
F_simpson = np.zeros(len(x_vals))
F_segmented_simpson = np.zeros(len(x_vals))
F_segmented_trapz = np.zeros(len(x_vals))
F_gauss_legendre = np.zeros(len(x_vals))

# 自定义 cumtrapz 实现（使用 numpy 的 cumsum）
def custom_cumtrapz(y, x):
    dx = np.diff(x)
    return np.concatenate([[0], cumsum(dx * (y[:-1] + y[1:]) / 2)])

# 简单梯形方法：在整个区间 [0, xi] 上计算梯形积分
def simple_trapz(y, x):
    h = (x[-1] - x[0])
    return h * (y[0] + y[-1]) / 2

# 简单 Simpson 方法：在整个区间 [0, xi] 上计算 Simpson 积分
def simple_simpson(y, x):
    h = (x[-1] - x[0]) / 2
    return (h / 3) * (y[0] + 4 * y[1] + y[2])

# 计算积分
for i, xi in enumerate(x_vals):
    sub_x = np.array([0, xi / 2, xi])  # Simpson 方法需要三个点
    sub_f = f(sub_x)

    # SciPy 的 quad 方法 (相当于 MATLAB 的 integral)
    F_integral[i], _ = quad(f, 0, xi)

    # 简单梯形公式
    F_custom_trapz[i] = simple_trapz(sub_f, sub_x)

    # NumPy 的 trapz 方法
    F_scipy_trapz[i] = trapz(sub_f, sub_x)

    # 自定义 cumtrapz 实现
    sub_x_full = np.linspace(0, xi, n_segments + 1)
    sub_f_full = f(sub_x_full)
    F_cumtrapz_vals = custom_cumtrapz(sub_f_full, sub_x_full)
    F_cumtrapz[i] = F_cumtrapz_vals[-1]

    # 简单 Simpson 方法
    F_simpson[i] = simple_simpson(sub_f, sub_x)

    # 自定义分段 Simpson 方法（复合 Simpson 仅作为对比）
    F_segmented_simpson[i] = 0
    for j in range(0, n_segments, 2):
        F_segmented_simpson[i] += (xi / n_segments / 3) * (sub_f_full[j] + 4 * sub_f_full[j + 1] + sub_f_full[j + 2])

    # 自定义分段梯形方法（复合梯形仅作为对比）
    F_segmented_trapz[i] = 0
    for j in range(n_segments):
        F_segmented_trapz[i] += (xi / n_segments / 2) * (sub_f_full[j] + sub_f_full[j + 1])

    # SciPy 的 Gauss-Legendre (quadrature 方法)
    F_gauss_legendre[i], _ = quadrature(f, 0, xi)

# 计算每种方法的误差 (二范数)
error_integral = np.linalg.norm(F_integral - F_exact)
error_custom_trapz = np.linalg.norm(F_custom_trapz - F_exact)
error_scipy_trapz = np.linalg.norm(F_scipy_trapz - F_exact)
error_cumtrapz = np.linalg.norm(F_cumtrapz - F_exact)
error_simpson = np.linalg.norm(F_simpson - F_exact)
error_segmented_simpson = np.linalg.norm(F_segmented_simpson - F_exact)
error_segmented_trapz = np.linalg.norm(F_segmented_trapz - F_exact)
error_gauss_legendre = np.linalg.norm(F_gauss_legendre - F_exact)

# 分类输出误差
print('--- SciPy 的 integral 方法 (quad) ---')
print(f'Integral 方法误差（二范数）：{error_integral:.10f}')

print('\n--- 简单梯形方法 ---')
print(f'简单梯形公式误差（二范数）：{error_custom_trapz:.10f}')
print(f'NumPy trapz 方法误差（二范数）：{error_scipy_trapz:.10f}')
print(f'自定义 cumtrapz 方法误差（二范数）：{error_cumtrapz:.10f}')
print(f'分段复合梯形方法误差（二范数）：{error_segmented_trapz:.10f}')

print('\n--- 简单 Simpson 方法 ---')
print(f'简单 Simpson 方法误差（二范数）：{error_simpson:.10f}')
print(f'分段复合 Simpson 方法误差（二范数）：{error_segmented_simpson:.10f}')

print('\n--- 高斯-勒让德方法 ---')
print(f'Gauss-Legendre 方法误差（二范数）：{error_gauss_legendre:.10f}')

# 绘制积分结果与原函数的比较
plt.figure()
plt.plot(x_vals, F_exact, 'k-', linewidth=2, label='原函数 y = sin(πx)')
plt.plot(x_vals, F_integral, 'b-', linewidth=1.5, label='Integral 方法 (quad)')
plt.plot(x_vals, F_custom_trapz, 'g--', linewidth=1.5, label='简单梯形公式')
plt.plot(x_vals, F_scipy_trapz, 'y-.', linewidth=1.5, label='NumPy trapz 方法')
plt.plot(x_vals, F_cumtrapz, 'c-', linewidth=1.5, label='自定义 cumtrapz 方法')
plt.plot(x_vals, F_segmented_trapz, 'b--', linewidth=1.5, label='分段复合梯形')
plt.plot(x_vals, F_simpson, 'm-.', linewidth=1.5, label='简单 Simpson 方法')
plt.plot(x_vals, F_segmented_simpson, 'r-', linewidth=1.5, label='分段 Simpson')
plt.plot(x_vals, F_gauss_legendre, 'c:', linewidth=1.5, label='Gauss-Legendre 方法 (quadrature)')
plt.title('数值积分方法与原函数 y = sin(πx) 的比较')
plt.xlabel('x')
plt.ylabel('F(x)')
plt.legend(loc='best')
plt.grid()
plt.show()

# 绘制误差比较
plt.figure()
plt.plot(x_vals, np.abs(F_integral - F_exact), 'b-', linewidth=1.5, label='Integral 方法 (quad)')
plt.plot(x_vals, np.abs(F_custom_trapz - F_exact), 'g--', linewidth=1.5, label='简单梯形公式')
plt.plot(x_vals, np.abs(F_scipy_trapz - F_exact), 'y-.', linewidth=1.5, label='NumPy trapz 方法')
plt.plot(x_vals, np.abs(F_cumtrapz - F_exact), 'c-', linewidth=1.5, label='自定义 cumtrapz 方法')
plt.plot(x_vals, np.abs(F_segmented_trapz - F_exact), 'b--', linewidth=1.5, label='分段复合梯形')
plt.plot(x_vals, np.abs(F_simpson - F_exact), 'm-.', linewidth=1.5, label='简单 Simpson 方法')
plt.plot(x_vals, np.abs(F_segmented_simpson - F_exact), 'r-', linewidth=1.5, label='分段 Simpson')
plt.plot(x_vals, np.abs(F_gauss_legendre - F_exact), 'c:', linewidth=1.5, label='Gauss-Legendre 方法 (quadrature)')
plt.title('数值积分方法的误差比较')
plt.xlabel('x')
plt.ylabel('误差 |F_numerical - F_exact|')
plt.legend(loc='best')
plt.grid()
plt.show()

File: integral_test.m
close all; % 关闭所有图形窗口

% 参数设置：积分区间分段次数（必须是偶数，以满足 Simpson 公式的要求）
n_segments = 4; % 可以调整此参数控制分段次数

% 定义被积函数
f = @(x) pi * cos(pi * x);

% 定义积分区间
a = -1;
b = 1;

% 定义离散点
x_vals = linspace(a, b, 100); % 主要用于绘制的采样点

% 计算原函数 F(x) = sin(pi * x)
F_exact = sin(pi * x_vals);

% 初始化用于存储每种方法的积分结果
F_integral = zeros(size(x_vals));
F_matlab_trapz = zeros(size(x_vals));
F_custom_trapz = zeros(size(x_vals));
F_cumtrapz = zeros(size(x_vals));
F_simpson = zeros(size(x_vals));
F_segmented_simpson = zeros(size(x_vals));
F_segmented_trapz = zeros(size(x_vals));
F_gauss_legendre = zeros(size(x_vals));

% 循环计算每个点的积分
for i = 1:length(x_vals)
    xi = x_vals(i);
    sub_x = linspace(0, xi, n_segments + 1); % 将 [0, xi] 分为 n_segments 等分
    sub_f = f(sub_x);
    h = (xi - 0) / n_segments; % 子区间的宽度

    % MATLAB 的 integral 方法
    F_integral(i) = integral(f, 0, xi); 

    % 梯形方法
    F_custom_trapz(i) = h * (sum(sub_f) - (sub_f(1) + sub_f(end)) / 2); % 自定义梯形公式
    F_matlab_trapz(i) = trapz(sub_x, sub_f);    % MATLAB 自带 trapz 方法
    F_cumtrapz_vals = cumtrapz(sub_x, sub_f);   % MATLAB 的 cumtrapz 方法
    F_cumtrapz(i) = F_cumtrapz_vals(end);       % 取 cumtrapz 的最终积分值
    F_segmented_trapz(i) = 0;  % 分段复合梯形
    for j = 1:n_segments
        F_segmented_trapz(i) = F_segmented_trapz(i) + (h / 2) * (sub_f(j) + sub_f(j+1));
    end

    % Simpson 方法
    F_simpson(i) = (h / 3) * (sub_f(1) + 4 * sum(sub_f(2:2:end-1)) + 2 * sum(sub_f(3:2:end-2)) + sub_f(end)); % 自定义 Simpson
    F_segmented_simpson(i) = 0;  % 分段复合 Simpson
    for j = 1:2:n_segments-1
        F_segmented_simpson(i) = F_segmented_simpson(i) + (h/3) * (sub_f(j) + 4*sub_f(j+1) + sub_f(j+2));
    end

    % 高斯-勒让德积分（quadgk）
    F_gauss_legendre(i) = quadgk(f, 0, xi);
end

% 计算每种方法相对于原函数的误差（使用二范数）
error_integral = norm(F_integral - F_exact, 2);
error_matlab_trapz = norm(F_matlab_trapz - F_exact, 2);
error_custom_trapz = norm(F_custom_trapz - F_exact, 2);
error_cumtrapz = norm(F_cumtrapz - F_exact, 2);
error_simpson = norm(F_simpson - F_exact, 2);
error_segmented_simpson = norm(F_segmented_simpson - F_exact, 2);
error_segmented_trapz = norm(F_segmented_trapz - F_exact, 2);
error_gauss_legendre = norm(F_gauss_legendre - F_exact, 2);

% 分类输出误差
fprintf('--- MATLAB 的 integral 方法 ---\n');
fprintf('Integral 方法误差（二范数）：%.10f\n', error_integral);

fprintf('\n--- 梯形方法 ---\n');
fprintf('自定义梯形公式误差（二范数）：%.10f\n', error_custom_trapz);
fprintf('MATLAB trapz 方法误差（二范数）：%.10f\n', error_matlab_trapz);
fprintf('MATLAB cumtrapz 方法误差（二范数）：%.10f\n', error_cumtrapz);
fprintf('分段复合梯形方法误差（二范数）：%.10f\n', error_segmented_trapz);

fprintf('\n--- Simpson 方法 ---\n');
fprintf('自定义 Simpson 方法误差（二范数）：%.10f\n', error_simpson);
fprintf('分段复合 Simpson 方法误差（二范数）：%.10f\n', error_segmented_simpson);

fprintf('\n--- MATLAB 的 高斯-勒让德方法 ---\n');
fprintf('Gauss-Legendre 方法误差（二范数）：%.10f\n', error_gauss_legendre);

% 绘制数值积分结果与原函数的比较
figure;
plot(x_vals, F_exact, 'k-', 'LineWidth', 2); % 原函数
hold on;
plot(x_vals, F_integral, 'b-', 'LineWidth', 1.5); % integral 方法
plot(x_vals, F_custom_trapz, 'g--', 'LineWidth', 1.5); % 自定义梯形公式
plot(x_vals, F_matlab_trapz, 'y-.', 'LineWidth', 1.5); % MATLAB 自带 trapz 方法
plot(x_vals, F_cumtrapz, 'c-', 'LineWidth', 1.5); % MATLAB cumtrapz 方法
plot(x_vals, F_segmented_trapz, 'b--', 'LineWidth', 1.5); % 分段复合梯形
plot(x_vals, F_simpson, 'm-.', 'LineWidth', 1.5); % Simpson 方法
plot(x_vals, F_segmented_simpson, 'r-', 'LineWidth', 1.5); % 分段 Simpson
plot(x_vals, F_gauss_legendre, 'c:', 'LineWidth', 1.5); % Gauss-Legendre 方法 (quadgk)

title('数值积分方法与原函数 y = \sin(\pi \cdot x) 的比较');
xlabel('x');
ylabel('F(x)');
legend('原函数 y = \sin(\pi \cdot x)', 'integral 方法', '自定义梯形公式', ...
       'MATLAB trapz 方法', 'MATLAB cumtrapz 方法', '分段复合梯形', ...
       'Simpson 方法', '分段 Simpson', 'Gauss-Legendre 方法', 'Location', 'Best');
grid on;
hold off;

% 绘制误差比较（用于可视化每种方法的相对误差）
figure;
plot(x_vals, abs(F_integral - F_exact), 'b-', 'LineWidth', 1.5);
hold on;
plot(x_vals, abs(F_custom_trapz - F_exact), 'g--', 'LineWidth', 1.5);
plot(x_vals, abs(F_matlab_trapz - F_exact), 'y-.', 'LineWidth', 1.5);
plot(x_vals, abs(F_cumtrapz - F_exact), 'c-', 'LineWidth', 1.5);
plot(x_vals, abs(F_segmented_trapz - F_exact), 'b--', 'LineWidth', 1.5);
plot(x_vals, abs(F_simpson - F_exact), 'm-.', 'LineWidth', 1.5);
plot(x_vals, abs(F_segmented_simpson - F_exact), 'r-', 'LineWidth', 1.5);
plot(x_vals, abs(F_gauss_legendre - F_exact), 'c:', 'LineWidth', 1.5);

title('数值积分方法的误差比较');
xlabel('x');
ylabel('误差 |F_{numerical} - F_{exact}|');
legend('integral 方法误差', '自定义梯形公式误差', 'MATLAB trapz 方法误差', ...
       'MATLAB cumtrapz 方法误差', '分段复合梯形方法误差', 'Simpson 方法误差', ...
       '分段 Simpson 方法误差', 'Gauss-Legendre 方法误差', 'Location', 'Best');
grid on;
hold off;
